zip-0311.rst

```diff
@@ -0,0 +1,111 @@
+  A Merkle mountain range (MMR) is binary hash tree that allows for efficient appends of new leaves without changing the value of existing nodes [#PeterTodd]_.
+
+Hard fork 
+  A hard fork is a change to a protocol that makes transactions from older clients invalid. 
+  
+
+Abstract
+========
+A proposed implementation of the probabilistic verification FlyClient protocol [#FlyClient]_ on top of the existing ZCash reference light client protocol [#ZIPXXX]_. This can be implemented with a hard fork to include a MMR proof in each block header.
+
+Motivation
+==========
+FlyClient reduces the number of block headers needed for light client verification of a valid chain, from linear (as in the current reference protocol) to logarithmic in blockchain length. This verification is correct with high probability. It also allows creation of subchain proofs, so light clients need only check blocks later than the most recently verified block index. Following that, verification of a transaction inclusion within that block follows the usual reference protocol [#ZIPXXX]_. 
+
+A smaller proof size could facilitate ‘side-chaining’ with other blockchains such as Ethereum. It also reduces bandwidth and storage requirements for limited clients like mobile or IoT devices.
+
+
+Specification
+=============
+
+Probabilistic sampling
+```````````````````````
+Instead of downloading every block header, the FlyClient samples blocks from the full node with the given sampling distribution. Here, x is the relative aggregate weight at the given block (e.g. if x=1/2 at some block, it means that half the chain's difficulty has been amassed in the interval up to that block), and δ is the fraction of the blocks which are always checked by the client with probability 1.
+
+.. figure:: pdf.png
+    :align: center
+    :scale: 50 %
+    :figclass: align-center
+    
+It has been found that as long as log_c(δ) is constant, where c is the adversary's mining power as a fraction of the honest mining power, then the number of light client queries needed are logarithmic in the size of the blockchain [#FlyClient]_.
+
+
+New block header
+`````````````````
+Each FlyClient prover has to maintain a commitment to all its block headers in a Merkle mountain range (MMR). This means that each ZCash block header must include the MMR root of all the blocks before it. This can be implemented in a hard fork of the ZCash protocol, in which the MMR root is added to the header of all old and new blocks. 
+
+Merkle mountain range
+``````````````````````
+
+.. figure:: MMR.png
+    :align: center
+    :figclass: align-center
+
+    Schematic of Merkle mountain range (from [#FlyClient]_).
+    
+The MMR structure allows for subchain proofs. Consider a client who has already received a verified a proof for a chain of n blocks. When it later receives k new blocks, it need only verify that the new section of length k is honest, as well as that the old chain is a prefix of the new chain.
```

Comment:
Maybe this is an aesthetic difference between the zcash and ethereum communities but it would be nice to see a formal description of exactly what this entails. I don't necessarily mean Latex or Coq, just a python function `def get_mmr_values(prev_mmr_values: List[Bytes32], block_hash: Bytes32) -> Bytes32` with the code. Maybe this code that I wrote a few months ago does the same thing? https://github.com/ethereum/research/blob/master/beacon_chain_impl/progressive_merkle_tree.py#L12

Reply:
In general: it's only strictly necessary to specify the rules necessary for conformance. (Another example of this is that the protocol spec specifies Equihash validation, but not how to construct Equihash proofs.) There's no objection to discussing anything else that is useful, though. Thanks for the code reference.